File: src/com/leozhi/topic222/Solution04.kt
package com.leozhi.topic222

/**
 * @author leozhi
 * 递归
 * 通过
 * 220ms
 */
class Solution04 {
    private fun countNodes(root: TreeNode?): Int {
        root ?: return 0
        val leftHeight = getHeight(root.left)
        val rightHeight = getHeight(root.right)
        if (leftHeight == rightHeight) {
            return getCount(root.left, 1) + 1 + countNodes(root.right)
        }
        return 1 + getCount(root.right, 1) + countNodes(root.left)
    }

    private fun getCount(root: TreeNode?, depth: Int): Int {
        root ?: return 0
        return depth + getCount(root.left, depth * 2)
    }

    private fun getHeight(root: TreeNode?): Int {
        root ?: return 0
        return 1 + getHeight(root.left)
    }
}

File: src/com/leozhi/topic605/Solution04.cpp
#include <vector>
using namespace std;

/**
 * @author leozhi
 * 贪心算法
 * 通过
 * 44ms
 */
class Solution {
public:
    bool canPlaceFlowers(vector<int>& flowerbed, int n) {
        if (n == 0) {
            return true;
        }
        int len = flowerbed.size();
        for (int i = 0; i < len; ++i) {
            bool left = i == 0 || flowerbed[i - 1] == 0;
            bool right = i == len - 1 || flowerbed[i + 1] == 0;
            if (flowerbed[i] == 0 && left && right) {
                flowerbed[i] = 1;
                if (--n == 0) {
                    return true;
                }
            }
        }
        return false;
    }
};